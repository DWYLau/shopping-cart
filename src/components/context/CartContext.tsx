import React, { createContext, useContext, useState, ReactNode } from "react"
import { Product } from "../../utils/types"

type CartContextType = {
  cart: Product[] | null
  setCart: React.Dispatch<React.SetStateAction<Product[] | null>>
}

type CartProviderProps = {
  children: ReactNode
}

const CartContext = createContext<CartContextType>({
  cart: [],
  setCart: () => {},
})

function useCart(): CartContextType {
  return useContext(CartContext)
}

function CartProvider({ children }: CartProviderProps): JSX.Element {
  const [cart, setCart] = useState<Product[] | null>(null)

  return (
    <CartContext.Provider value={{ cart, setCart }}>
      {children}
    </CartContext.Provider>
  )
}

export { useCart, CartProvider }
