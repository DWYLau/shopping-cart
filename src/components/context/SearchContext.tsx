import React, { createContext, useContext, useState, ReactNode } from "react"

type SearchContextType = {
  searchValue: string | null
  setSearchValue: React.Dispatch<React.SetStateAction<string | null>>
}

const SearchContext = createContext<SearchContextType>({
  searchValue: "",
  setSearchValue: () => {},
})

const useSearch = (): SearchContextType => useContext(SearchContext)

type SearchProviderProps = {
  children: ReactNode
}

const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [searchValue, setSearchValue] = useState<string | null>("")

  return (
    <SearchContext.Provider value={{ searchValue, setSearchValue }}>
      {children}
    </SearchContext.Provider>
  )
}

export { useSearch, SearchProvider }
