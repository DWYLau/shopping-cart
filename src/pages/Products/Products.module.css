.flex-container {
  display: flex;
  gap: 1rem;
}

.grid-container {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  width: 100%;
  row-gap: 1rem;
  margin-left: 1rem;
  margin-top: 3rem;
  margin-bottom: 2rem;
  align-items: center;
  justify-items: center;
}
