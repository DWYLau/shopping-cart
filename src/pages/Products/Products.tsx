import { useEffect, useState } from "react"
import { fetchData } from "../../utils/api"
import Sidebar from "../../components/Sidebar/Sidebar"
import styles from "./Products.module.css"

function Products() {
  const [products, setProducts] = useState<any>([])
  const [error, setError] = useState(null)
  const [loading, setLoading] = useState(true)

  useEffect(() => {
    {
      fetchData()
        .then(data => {
          setProducts(data)
          console.log(data)
        })
        .catch(error => setError(error))
        .finally(() => setLoading(false))
    }
  }, [])

  if (loading) return <div>Loading</div>

  if (!loading && products)
    return (
      <section className={styles["flex-container"]}>
        <Sidebar />
        <div className={styles["grid-container"]}>
          {products.map(product => {
            return (
              <div className={styles.card}>
                <img className={styles.image} src={product.image} alt='' />
                <h3>{product.title}</h3>
                <div>
                  <h3>{product.rating.rate}</h3>
                  <h3>{product.rating.count}</h3>
                </div>

                <h3>{product.price}</h3>
              </div>
            )
          })}
        </div>
      </section>
    )
}

export default Products
