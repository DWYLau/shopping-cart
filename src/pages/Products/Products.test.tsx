import { render } from "@testing-library/react"
import Products from "./Products"
import { describe, it, expect } from "vitest"
import { SearchProvider } from "../../components/context/SearchContext"
import { CartProvider } from "../../components/context/CartContext"

describe("Product component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <CartProvider>
        <SearchProvider>
          <Products />
        </SearchProvider>
      </CartProvider>
    )

    expect(container).toMatchSnapshot()
  })

  it("renders load correctly", () => {
    const { container } = render(
      <CartProvider>
        <SearchProvider>
          <Products />
        </SearchProvider>
      </CartProvider>
    )

    expect(container.querySelector(".loader")).toBeNull()
  })
})
