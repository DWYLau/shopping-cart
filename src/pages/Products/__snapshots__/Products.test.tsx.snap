// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product component > renders correctly 1`] = `
<div>
  <div
    class="_container_eb1f5f"
  >
    <div
      class="_loader_eb1f5f"
    />
  </div>
</div>
`;
